hobo/src/storage.rs

```diff
@@ -151,34 +154,91 @@ impl<Component: 'static> Storage<Component> for SimpleStorage<Component> {
+}
+
+unsafe impl<Component: 'static, Inner: std::ops::Deref<Target = SimpleStorage<Component>>> owning_ref::StableAddress for StorageGuard<Component, Inner> {}
+unsafe impl<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>> owning_ref::StableAddress for StorageGuardMut<'a, Component, Inner> {}
+
+impl<Component, Inner> std::ops::Deref for StorageGuard<Component, Inner> where
+	Component: 'static,
+	Inner: std::ops::Deref<Target = SimpleStorage<Component>>,
+{
+	type Target = SimpleStorage<Component>;
 
-pub struct StorageGuard<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>>(pub &'a World, pub Option<Inner>);
-unsafe impl<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>> owning_ref::StableAddress for StorageGuard<'a, Component, Inner> {}
+	fn deref(&self) -> &Self::Target { self.inner.as_ref().unwrap() }
+}
+
+#[cfg(debug_assertions)]
+impl<Component, Inner> Drop for StorageGuard<Component, Inner> where
+	Component: 'static,
+	Inner: std::ops::Deref<Target = SimpleStorage<Component>>,
+{
+	fn drop(&mut self) {
+		let StorageGuard { location, .. } = self;
+		crate::backtrace::STORAGE_MAP.0.borrow_mut()
+			.entry(std::any::TypeId::of::<Component>())
+			.and_modify(|map| { map.remove(location); });
+	}
+}
 
-impl<'a, Component, Inner> std::ops::Deref for StorageGuard<'a, Component, Inner> where
+impl<'a, Component, Inner> std::ops::Deref for StorageGuardMut<'a, Component, Inner> where
 	Component: 'static,
 	Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>,
 {
 	type Target = SimpleStorage<Component>;
 
-	fn deref(&self) -> &Self::Target { self.1.as_ref().unwrap() }
+	fn deref(&self) -> &Self::Target { self.inner.as_ref().unwrap() }
 }
 
-impl<'a, Component, Inner> std::ops::DerefMut for StorageGuard<'a, Component, Inner> where
+impl<'a, Component, Inner> std::ops::DerefMut for StorageGuardMut<'a, Component, Inner> where
 	Component: 'static,
 	Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>,
 {
-	fn deref_mut(&mut self) -> &mut Self::Target { self.1.as_mut().unwrap() }
+	fn deref_mut(&mut self) -> &mut Self::Target { self.inner.as_mut().unwrap() }
 }
 
 // dropping StorageGuard should trigger updates of relevant systems
 // right now it's pooling all entities that were involved in changes, additions or removals
-impl<'a, Component, Inner> Drop for StorageGuard<'a, Component, Inner> where
+impl<'a, Component, Inner> Drop for StorageGuardMut<'a, Component, Inner> where
 	Component: 'static,
 	Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>,
 {
+	#[cfg(debug_assertions)]
+	fn drop(&mut self) {
+		let StorageGuardMut { world, inner, location } = self;
+
+		let type_id = std::any::TypeId::of::<Component>();
+		let type_name = std::any::type_name::<Component>().to_owned();
+		crate::backtrace::STORAGE_MAP.0.borrow_mut()
+			.entry(type_id)
+			.and_modify(|map| { 
+				if map.len() > 1 {
```

Comment:
this is cleaner written as `assert_eq!(map.len(), 1, "...")`

asserts literally compile down to this, except with some extra